pkg/controller/staleness/staleness.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package staleness
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/resourceversion"
+	clientset "k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	compbasemetrics "k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/klog/v2"
+)
+
+var (
+	probePeriod              = 5 * time.Second
+	maxDelayMeasured         = 5 * time.Minute
+	queueSize                = int(maxDelayMeasured/probePeriod) + 1
+	minSample                = probePeriod / 2
+	bucketCount              = 10
+	bucketsExponentialFactor = math.Pow(float64(maxDelayMeasured)/float64(minSample), 1.0/float64(bucketCount))
+	watchDelayHistogram      = compbasemetrics.NewHistogramVec(
+		&compbasemetrics.HistogramOpts{
+			Subsystem:      "controller_manager",
+			Name:           "watch_delay_seconds",
+			Help:           "Watch delay seconds, for now calculated only for pods",
+			StabilityLevel: compbasemetrics.ALPHA,
+			Buckets:        compbasemetrics.ExponentialBuckets(minSample.Seconds(), bucketsExponentialFactor, bucketCount),
+		},
+		[]string{"group", "resource"},
+	)
+	watchDelayGauge = compbasemetrics.NewGaugeVec(
+		&compbasemetrics.GaugeOpts{
+			Subsystem:      "controller_manager",
+			Name:           "pod_watch_delay_seconds",
+			Help:           "Pod watch delay seconds",
+			StabilityLevel: compbasemetrics.ALPHA,
+		},
+		[]string{"group", "resource"},
+	)
+)
+
+func init() {
+	legacyregistry.MustRegister(watchDelayGauge)
+	legacyregistry.MustRegister(watchDelayHistogram)
+}
+
+type RVTime struct {
+	rv  string
+	now time.Time
+}
+
+type PodProber struct {
+	logger klog.Logger
+	// client interface
+	kubeClient clientset.Interface
+	// podLister can list/get pods from the shared informers's store
+	podLister corelisters.PodLister
+
+	mux     sync.Mutex
+	rvQueue []RVTime
+	curRV   string
+}
+
+func (p *PodProber) runProbeLoop(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(probePeriod):
+		}
+		start := time.Now()
+		resp, err := p.kubeClient.CoreV1().Pods("").List(ctx, metav1.ListOptions{})
```

Comment:
Will this not return all pods? What's the result of passing "" to namespace, I would guess it's defaulted to "default"